### PR TITLE
subset: Add grammar and kind

### PR DESCRIPTION
**Pull request by jcbrinfo**
Implement the syntactic, inheritance and instantiation rules.

Obsoletes: #2469

### CODE REVIEW

**Comment by jcbrinfo** [comment]
`tests_are_ok`. Only `test_neo4j_batch` failed.

**Comment by jcbrinfo** [comment]
♻️ 
- [x] Split changes to `MClassKind::can_specialize` (as @privat suggested)

**Comment by jcbrinfo** [comment]
♻️ on top of `master`

**Comment by jcbrinfo** [comment]
♻️ 
- [x] Deny signature definitions for predicates.

File: src/model/model.nit [hunk 1]
```diff
@@ -2602,12 +2602,10 @@ class MClassKind
 		else if self == extern_kind then
 			# only compatible with themselves
 			return self == other
-		else if other == enum_kind or other == extern_kind then
-			# abstract_kind and concrete_kind are incompatible
-			return false
+		else
+			# assert self == abstract_kind or self == concrete_kind
```
Comment (privat) [inline src/model/model.nit]: this comment is weird

Comment (jcbrinfo) [inline src/model/model.nit]: It is unambiguous at least.

Do you want that I uncomment that line?

Comment (privat) [inline src/model/model.nit]: it's ok

**Pull request merged**